File: docs/src/main/java/jexamples/virtualnetworking/pingpongselectors/PingPongSerializer.java
package jexamples.virtualnetworking.pingpongselectors;

import java.util.Optional;
import io.netty.buffer.ByteBuf;
import se.sics.kompics.network.netty.serialization.Serializer;

public class PingPongSerializer implements Serializer {

  private static final byte PING = 1;
  private static final byte PONG = 2;

  @Override
  public int identifier() {
    return 200;
  }

  @Override
  public void toBinary(Object o, ByteBuf buf) {
    if (o instanceof Ping) {
      buf.writeByte(PING); // 1 byte
      // total 1 bytes
    } else if (o instanceof Pong) {
      buf.writeByte(PONG); // 1 byte
      // total 1 bytes
    }
  }

  @Override
  public Object fromBinary(ByteBuf buf, Optional<Object> hint) {
    byte type = buf.readByte(); // 1 byte
    switch (type) {
      case PING:
        return Ping.EVENT; // 1 bytes total, check
      case PONG:
        return Pong.EVENT; // 1 bytes total, check
    }
    return null;
  }
}

File: docs/src/main/scala/sexamples/virtualnetworking/pingpongvirtual/PingPongSerializer.scala
package sexamples.virtualnetworking.pingpongvirtual

import io.netty.buffer.ByteBuf
import java.net.{InetAddress, UnknownHostException}
import java.util.Optional
import se.sics.kompics.network.Transport
import se.sics.kompics.network.netty.serialization.{Serializer, Serializers}

object PingPongSerializer extends Serializer {
  private val PING: Byte = 1;
  private val PONG: Byte = 2;

  private val NO_HINT: Optional[AnyRef] = Optional.empty();

  override def identifier(): Int = 200;

  override def toBinary(o: AnyRef, buf: ByteBuf): Unit = {
    o match {
      case Ping => {
        buf.writeByte(PING); // 1 byte
        // total 1 bytes
      }
      case Pong => {
        buf.writeByte(PONG); // 1 byte
        // total 1 bytes
      }
    }
  }

  override def fromBinary(buf: ByteBuf, hint: Optional[AnyRef]): AnyRef = {
    val typeFlag = buf.readByte(); // 1 byte
    typeFlag match {
      case PING => {
        return Ping; // 1 bytes total, check
      }
      case PONG => {
        return Pong; // 1 bytes total, check
      }
      case _ => {
        Console.err.println(s"Got invalid byte flag=$typeFlag");
        return null;
      }
    }
  }
}
